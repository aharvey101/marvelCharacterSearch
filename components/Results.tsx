import {
  Button,
  Alert,
  Card,
  Col,
  Row,
  Spinner,
  Container,
  Modal,
} from 'react-bootstrap'
import Image from 'next/image'
import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { ResultsEntity } from '../types'

const StyledCard = styled(Card)`
  margin-top: 2rem;
  width: 18rem;
  height: 90%;
`

const StyledSpacer = styled.div`
  margin: 2rem;
`

interface ICards {
  characters: ResultsEntity[]
  isLoading: boolean
}

interface ICharacterCard {
  image: string
  name: string
  setShowModal: (id: number) => void
  showMore: boolean
  id: number
}

interface ICustomModal {
  image: string
  name: string
  description: string
  modalState: boolean
  id: number
  handleModal: (id: number) => void
}

interface IModalState {
  id: number
  modalState: boolean
}

const CharacterCard: React.FC<ICharacterCard> = ({
  name,
  image,
  setShowModal,
  showMore,
  id,
}) => {
  return (
    <Row>
      <Col>
        <StyledCard>
          <Card.Img variant="top" src={image} />
          <Card.Body>
            <Card.Title>{name}</Card.Title>
            {showMore ? (
              <Button onClick={() => setShowModal(id)} variant="primary">
                Show More
              </Button>
            ) : null}
          </Card.Body>
        </StyledCard>
      </Col>
    </Row>
  )
}

export const Cards: React.FC<ICards> = ({ characters, isLoading }): any => {
  const [modalStateArray, setModalStateArray] = useState<IModalState[]>([])

  const MyModal: React.FC<ICustomModal> = ({
    image,
    name,
    description,
    modalState,
    id,
    handleModal,
  }) => {
    return (
      <Modal show={modalState} backdrop="static" keyboard={false}>
        <Modal.Header>
          <Modal.Title>{name}</Modal.Title>
        </Modal.Header>
        <Image height={270} width={200} src={image} alt={name}></Image>
        <Modal.Body>{description}</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={() => handleModal(id)}>
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    )
  }

  useEffect(() => {
    if (characters) {
      const modalStates = characters.map((character) => {
        return {
          id: character.id,
          modalState: false,
        }
      })
      setModalStateArray(modalStates)
    }
  }, [characters])

  const handleModal = (id: number) => {
    const updatedEntry: IModalState | undefined = modalStateArray.find(
      (modalState) => modalState.id === id
    )
    if (updatedEntry) {
      updatedEntry.modalState = !updatedEntry.modalState
      setModalStateArray((ps) => [...ps, updatedEntry])
    }
  }

  const getModalState = (id: number) => {
    const state = modalStateArray.find((modal) => modal.id === id)
    if (state) {
      return state.modalState
    }
    return false
  }

  if (isLoading)
    return (
      <StyledSpacer>
        <Row className="justify-content-center">
          <Col />
          <Col className="d-flex justify-content-center">
            <Spinner animation="border" variant="warning"></Spinner>
          </Col>
          <Col />
        </Row>
      </StyledSpacer>
    )

  if (characters && characters.length <= 0) {
    return (
      <StyledSpacer>
        <Row>
          <Col />
          <Col sm={12} md={8} className=" d-flex justify-content-center">
            <Alert variant="danger">
              <Alert.Heading>No character with that name</Alert.Heading>
            </Alert>
          </Col>
          <Col />
        </Row>
      </StyledSpacer>
    )
  }

  if (characters) {
    const CharacterCards = characters.map((character) => (
      <Col key={character.id} className=" d-flex justify-content-center">
        <MyModal
          modalState={getModalState(character.id)}
          name={character.name}
          description={character.description}
          image={`${character.thumbnail.path}/landscape_xlarge.${character.thumbnail.extension}`}
          id={character.id}
          handleModal={handleModal}
        />
        <CharacterCard
          setShowModal={handleModal}
          name={character.name}
          image={`${character.thumbnail.path}/landscape_xlarge.${character.thumbnail.extension}`}
          showMore={character.description === '' ? false : true}
          id={character.id}
        ></CharacterCard>
      </Col>
    ))
    return <Row>{CharacterCards}</Row>
  }
  return null
}
