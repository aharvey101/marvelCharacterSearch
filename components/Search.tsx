import React, { useState } from 'react'
import { Form, Button, Row, Col } from 'react-bootstrap'

interface ISearch {
  handleSearch: (searchParam: string) => void
}

// TODO: fix up styling
export const Search: React.FC<ISearch> = ({ handleSearch }) => {
  const [character, setCharacter] = useState('')

  const handleUpdateState = (e): void => {
    setCharacter(e.target.value)
  }

  const handleSubmit = (event) => {
    event.preventDefault()
    handleSearch(character)
  }
  return (
    <Form onSubmit={handleSubmit}>
      <Row className="justify-content-md-center">
        <Col xs lg={2} />
        <Col xs={6} lg={8} md="auto">
          <Form.Group>
            <Form.Control onChange={handleUpdateState} />
          </Form.Group>
        </Col>
        <Col xs lg={2}>
          <Button variant="primary" type="submit">
            Search
          </Button>
        </Col>
      </Row>
    </Form>
  )
}
