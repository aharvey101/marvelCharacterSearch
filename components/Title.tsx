import React from 'react'
import { StyledTitleContainer, StyledTitle } from './styles/Title'

export const Title: React.FC = () => {
  return (
    <StyledTitleContainer>
      <StyledTitle>Marvel Character Finder</StyledTitle>
    </StyledTitleContainer>
  )
}
