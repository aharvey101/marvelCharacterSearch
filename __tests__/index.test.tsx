import Enzyme, { mount, ReactWrapper } from 'enzyme'
import axios, { AxiosStatic } from 'axios'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import Home from '../pages/index'
import { ISearch } from '../components/Search'
import { Search } from '../components/Search'
import { results } from '../resultsExample'
import characterSearch from '../util/searchAPI'

Enzyme.configure({ adapter: new Adapter() })

const mockSearchProps: ISearch = {
  handleSearch: jest.fn(),
  suggestions: ['test', 'test2'],
}

const data = { data: { data: results.data.results } }
jest.mock('../util/searchAPI', () => ({
  characterSearch: jest.fn(
    () => new Promise((resolve) => resolve({ test: 'test' }))
  ),
}))

jest.mock('axios')
describe('Index', () => {
  let wrapper: ReactWrapper

  beforeEach(() => {
    jest.clearAllMocks()
    wrapper = mount(
      <Home>
        <Search {...mockSearchProps} />
      </Home>
    )
  })

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Does not call API when input is empty', () => {
    wrapper.find('input').simulate('change', { target: { value: '' } })
    wrapper.find('button').simulate('submit')
    expect(axios.get).toBeCalledTimes(0)
  })

  xit('Calls API and updates state', async () => {
    wrapper.find('input').simulate('change', { target: { value: 'spiderman' } })
    wrapper.find('button').simulate('submit')
    // expect().toHaveBeenCalled()
  })
})
