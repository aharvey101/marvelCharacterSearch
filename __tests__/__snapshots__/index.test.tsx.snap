// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders 1`] = `ReactWrapper {}`;

exports[`first renders 1`] = `ReactWrapper {}`;
